pkg/virt-launcher/virtwrap/network/podinterface.go

```diff
@@ -97,10 +98,33 @@ func newPodNIC(vmi *v1.VirtualMachineInstance, network *v1.Network, handler netd
 		podInterfaceName: podInterfaceName,
 		iface:            correspondingNetworkIface,
 		launcherPID:      launcherPID,
-		dhcpConfigurator: &dhcpConfigurator,
+		dhcpConfigurator: dhcpConfigurator,
 	}, nil
 }
 
+func newBridgeDhcpConfigurator(iface v1.Interface, vmi v1.VirtualMachineInstance, podInterfaceName string, launcherPID *int, handler netdriver.NetworkHandler, cacheFactory cache.InterfaceCacheFactory) *dhcpconfigurator.DhcpConfigurator {
```

Comment:
I find the code added in this commit over complex and overkill. I believe the amount of dhcp related code in `podinterface.go` wasn't reduced but 2 new files were added. Having `BridgeDhcpServer, MasqueradeDhcpServer,  BridgeDhcpConfigurator, BridgeMasqueradeDhcpConfigurator`, just to call one `handler.StartDHCP` method with the correct parameters seems like an overkill to me. I believe the code prior this commit was better.

I would just pass to the new `EnsureDhcp` the relevalt `binding.startDHCP` method.

Reply:
The previous code misleads any reader into thinking that:
  - a bind mechanism requires DHCP. They don't. Only a single one has DHCP, and another might have it, depending on the CNI configuration.
  - forces all bind mechanisms to implement 3 methods, while half of them don't use them. That alone is a clear indication the interface is not correctly designed.

While I'm not saying the code I'm proposing is perfect, it has clear advantages to the previous one that I think you forgot to mention, or didn't notice. Sure. It's more code. It takes more code to make things explicit.

I'm not using those 4 added files to "just to call one `handler.StartDHCP` method", but to call that method **only** when it is required, while making it explicit to anyone reading this code that macvtap and slirp have absolutely no dhcp requirements. 

If you think that the previous code is better, I just ask you to look at the unit tests of the pod interface, and to the unit tests that I'm delivering here. Which ones are easier to understand ? Which ones make the reader better understand the code ? I think my proposal reads a lot better.

While saying all that, if you insist, I can drop the `DhcpServer` interface, and just pass a function along. I think that's a mistake (the bridge DHCP server requires the VMI, while masquerade doesn't). But composing the bind implementations with a dhcp configurator makes the code a lot more readable and easier to reason about.

Reply:
I've tried to drop the `DhcpServer` interface and use a function instead, but our unit tests fail on the network tests (check the goveralls target); apparently the `consistsOf` / `ContainElements` matchers do not play along with function struct members.

I've opened https://github.com/onsi/gomega/issues/445 to track it.

I recommend - again - using interfaces instead of different functions. I do not think it makes it that more complicated. I can squash both files into the configurator if that somehow helps out.

**Note:** using the `equals` matcher works. I'm not against using it, since the 2 of the 3 failing tests have a single item on the slice, and the remaining test's order of the podNICs is consistent (it is built based on a slice, which has a notion of order).

One thing I can't for the life of me understand is that `make test` works (after bumping gomega to latest version), while `make goveralls` doesn't. Furthermore, I managed to reproduce the issue in latest gomega, as per the bug linked above.

Reply:
I'm now proposing an alternative that uses the same code for both (we just have to make sure the `dhcpConfig` object features the correct parameters, and it will ensure the desired behavior, using the same code).

I think this has a better chance of being accepted.

Reply:
Resolving this one as well, the new proposed approach makes the solution simpler, and as such, this no longer applies.

It is also so different that nothing in this discussion can relate to it.